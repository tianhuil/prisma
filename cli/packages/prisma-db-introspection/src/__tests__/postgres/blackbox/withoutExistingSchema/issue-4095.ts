import testSchema from '../common'

describe('Introspector', () => {
  test('issue4095 relations', async () => {
    await testSchema(`-- -------------------------------------------------------------
    -- TablePlus 1.5(190)
    --
    -- https://tableplus.com/
    --
    -- Database: issue4019
    -- Generation Time: 2019-02-08 19:22:20.9670
    -- -------------------------------------------------------------
    
    
    DROP TABLE IF EXISTS "billing"."BillingAccount";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."BillingAccount_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."BillingAccount" (
        "id" int8 NOT NULL DEFAULT nextval('billing."BillingAccount_id_seq"'::regclass),
        "name" varchar(50) NOT NULL,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        "customerId" int8 NOT NULL,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."Condition";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."Condition_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."Condition" (
        "id" int8 NOT NULL DEFAULT nextval('billing."Condition_id_seq"'::regclass),
        "startDate" timestamptz NOT NULL,
        "endDate" timestamptz NOT NULL,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        "productId" int8 NOT NULL,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."Contract";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."Contract_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."Contract" (
        "id" int8 NOT NULL DEFAULT nextval('billing."Contract_id_seq"'::regclass),
        "code" varchar(50) NOT NULL,
        "enabled" bool DEFAULT false,
        "activationDateTime" timestamptz,
        "minimumFee" numeric NOT NULL,
        "lookToBook" numeric NOT NULL,
        "customerId" int8 NOT NULL,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."Cost";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."Cost_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."Cost" (
        "id" int8 NOT NULL DEFAULT nextval('billing."Cost_id_seq"'::regclass),
        "type" varchar(50) NOT NULL,
        "min" int4,
        "max" int4,
        "amount" numeric,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        "contractId" int8,
        "itemId" int8,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."Customer";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."Customer_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."Customer" (
        "id" int8 NOT NULL DEFAULT nextval('billing."Customer_id_seq"'::regclass),
        "code" varchar(50) NOT NULL,
        "name" varchar(100) NOT NULL,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."GlobalCondition";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."GlobalCondition_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."GlobalCondition" (
        "id" int8 NOT NULL DEFAULT nextval('billing."GlobalCondition_id_seq"'::regclass),
        "type" varchar(50) NOT NULL,
        "numConnections" int4,
        "millionSearches" int4,
        "amount" numeric,
        "globalStartFreeTrialPeriod" timestamptz,
        "globalEndFreeTrialPeriod" timestamptz,
        "defaultFreeTrialDaysForItem" int4,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        "contractId" int8 NOT NULL,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."Invoice";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."Invoice_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."Invoice" (
        "id" int8 NOT NULL DEFAULT nextval('billing."Invoice_id_seq"'::regclass),
        "startDate" timestamptz NOT NULL,
        "endDate" timestamptz NOT NULL,
        "paymentMethod" varchar(100) NOT NULL,
        "paymentStatus" varchar(100) NOT NULL,
        "amount" numeric NOT NULL,
        "amountPaid" numeric NOT NULL,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        "billingAccountId" int8 NOT NULL,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."InvoiceBreakdown";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."InvoiceBreakdown_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."InvoiceBreakdown" (
        "id" int8 NOT NULL DEFAULT nextval('billing."InvoiceBreakdown_id_seq"'::regclass),
        "startDate" timestamptz NOT NULL,
        "endDate" timestamptz NOT NULL,
        "amount" numeric NOT NULL,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        "itemBillingAccountId" int8 NOT NULL,
        "invoiceId" int8 NOT NULL,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."Item";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."Item_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."Item" (
        "id" int8 NOT NULL DEFAULT nextval('billing."Item_id_seq"'::regclass),
        "enabled" bool DEFAULT false,
        "specificFreeTrialDays" int4,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        "contractId" int8 NOT NULL,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."Item_BillingAccount";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."Item_BillingAccount_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."Item_BillingAccount" (
        "id" int8 NOT NULL DEFAULT nextval('billing."Item_BillingAccount_id_seq"'::regclass),
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        "billingAccountId" int8 NOT NULL,
        "itemId" int8 NOT NULL,
        PRIMARY KEY ("id")
    );
    
    DROP TABLE IF EXISTS "billing"."Product";
    -- This script only contains the table creation statements and does not fully represent the table in the database. It's still missing: indices, triggers. Do not use it as a backup.
    
    -- Sequence and defined type
    CREATE SEQUENCE IF NOT EXISTS billing."Product_id_seq";
    
    -- Table Definition
    CREATE TABLE "billing"."Product" (
        "id" int8 NOT NULL DEFAULT nextval('billing."Product_id_seq"'::regclass),
        "type" varchar(50) NOT NULL,
        "system" varchar(200) NOT NULL,
        "createdAt" timestamptz NOT NULL,
        "updatedAt" timestamptz NOT NULL,
        "itemId" int8 NOT NULL,
        PRIMARY KEY ("id")
    );
    
    ALTER TABLE "billing"."BillingAccount" ADD FOREIGN KEY ("customerId") REFERENCES "billing"."Customer"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."Condition" ADD FOREIGN KEY ("productId") REFERENCES "billing"."Product"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."Contract" ADD FOREIGN KEY ("customerId") REFERENCES "billing"."Customer"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."Cost" ADD FOREIGN KEY ("contractId") REFERENCES "billing"."Contract"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."Cost" ADD FOREIGN KEY ("itemId") REFERENCES "billing"."Item"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."GlobalCondition" ADD FOREIGN KEY ("contractId") REFERENCES "billing"."Contract"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."Invoice" ADD FOREIGN KEY ("billingAccountId") REFERENCES "billing"."BillingAccount"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."InvoiceBreakdown" ADD FOREIGN KEY ("invoiceId") REFERENCES "billing"."Invoice"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."InvoiceBreakdown" ADD FOREIGN KEY ("itemBillingAccountId") REFERENCES "billing"."Item_BillingAccount"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."Item" ADD FOREIGN KEY ("contractId") REFERENCES "billing"."Contract"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."Item_BillingAccount" ADD FOREIGN KEY ("billingAccountId") REFERENCES "billing"."BillingAccount"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."Item_BillingAccount" ADD FOREIGN KEY ("itemId") REFERENCES "billing"."Item"("id") ON DELETE CASCADE ON UPDATE CASCADE;
    ALTER TABLE "billing"."Product" ADD FOREIGN KEY ("itemId") REFERENCES "billing"."Item"("id") ON DELETE CASCADE ON UPDATE CASCADE;`,
      'billing')
  })
})
