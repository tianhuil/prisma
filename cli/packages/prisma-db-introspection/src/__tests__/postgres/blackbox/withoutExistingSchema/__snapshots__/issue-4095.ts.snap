// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Introspector issue4095 relations 1`] = `
"type BillingAccount {
  id: Int! @unique
  createdAt: DateTime!
  customerId: Customer!
  invoice: [Invoice]
  itemBillingAccount: [ItemBillingAccount]
  name: String!
  updatedAt: DateTime!
}

type Condition {
  id: Int! @unique
  createdAt: DateTime!
  endDate: DateTime!
  productId: Product!
  startDate: DateTime!
  updatedAt: DateTime!
}

type Contract {
  id: Int! @unique
  activationDateTime: DateTime
  code: String!
  cost: [Cost]
  createdAt: DateTime!
  customerId: Customer!
  enabled: Boolean @default(value: false)
  globalCondition: [GlobalCondition]
  item: [Item]
  lookToBook: Float!
  minimumFee: Float!
  updatedAt: DateTime!
}

type Cost {
  id: Int! @unique
  amount: Float
  contractId: Contract
  createdAt: DateTime!
  itemId: Item
  max: Int
  min: Int
  type: String!
  updatedAt: DateTime!
}

type Customer {
  id: Int! @unique
  billingAccount: [BillingAccount]
  code: String!
  contract: [Contract]
  createdAt: DateTime!
  name: String!
  updatedAt: DateTime!
}

type GlobalCondition {
  id: Int! @unique
  amount: Float
  contractId: Contract!
  createdAt: DateTime!
  defaultFreeTrialDaysForItem: Int
  globalEndFreeTrialPeriod: DateTime
  globalStartFreeTrialPeriod: DateTime
  millionSearches: Int
  numConnections: Int
  type: String!
  updatedAt: DateTime!
}

type Invoice {
  id: Int! @unique
  amount: Float!
  amountPaid: Float!
  billingAccountId: BillingAccount!
  createdAt: DateTime!
  endDate: DateTime!
  invoiceBreakdown: [InvoiceBreakdown]
  paymentMethod: String!
  paymentStatus: String!
  startDate: DateTime!
  updatedAt: DateTime!
}

type InvoiceBreakdown {
  id: Int! @unique
  amount: Float!
  createdAt: DateTime!
  endDate: DateTime!
  invoiceId: Invoice!
  itemBillingAccountId: ItemBillingAccount!
  startDate: DateTime!
  updatedAt: DateTime!
}

type Item {
  id: Int! @unique
  contractId: Contract!
  cost: [Cost]
  createdAt: DateTime!
  enabled: Boolean @default(value: false)
  itemBillingAccount: [ItemBillingAccount]
  product: [Product]
  specificFreeTrialDays: Int
  updatedAt: DateTime!
}

type ItemBillingAccount @pgTable(name: \\"Item_BillingAccount\\") {
  id: Int! @unique
  billingAccountId: BillingAccount!
  createdAt: DateTime!
  invoiceBreakdown: [InvoiceBreakdown]
  itemId: Item!
  updatedAt: DateTime!
}

type Product {
  id: Int! @unique
  condition: [Condition]
  createdAt: DateTime!
  itemId: Item!
  system: String!
  type: String!
  updatedAt: DateTime!
}"
`;

exports[`Introspector issue4095 relations 2`] = `
"type BillingAccount {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"BillingAccount_id_seq\\", initialValue: 1, allocationSize: 1)
  createdAt: DateTime! @createdAt
  customerId: Customer!
  invoice: [Invoice]
  itemBillingAccount: [ItemBillingAccount]
  name: String!
  updatedAt: DateTime! @updatedAt
}

type Condition {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"Condition_id_seq\\", initialValue: 1, allocationSize: 1)
  createdAt: DateTime! @createdAt
  endDate: DateTime!
  productId: Product!
  startDate: DateTime!
  updatedAt: DateTime! @updatedAt
}

type Contract {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"Contract_id_seq\\", initialValue: 1, allocationSize: 1)
  activationDateTime: DateTime
  code: String!
  cost: [Cost]
  createdAt: DateTime! @createdAt
  customerId: Customer!
  enabled: Boolean @default(value: false)
  globalCondition: [GlobalCondition]
  item: [Item]
  lookToBook: Float!
  minimumFee: Float!
  updatedAt: DateTime! @updatedAt
}

type Cost {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"Cost_id_seq\\", initialValue: 1, allocationSize: 1)
  amount: Float
  contractId: Contract
  createdAt: DateTime! @createdAt
  itemId: Item
  max: Int
  min: Int
  type: String!
  updatedAt: DateTime! @updatedAt
}

type Customer {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"Customer_id_seq\\", initialValue: 1, allocationSize: 1)
  billingAccount: [BillingAccount]
  code: String!
  contract: [Contract]
  createdAt: DateTime! @createdAt
  name: String!
  updatedAt: DateTime! @updatedAt
}

type GlobalCondition {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"GlobalCondition_id_seq\\", initialValue: 1, allocationSize: 1)
  amount: Float
  contractId: Contract!
  createdAt: DateTime! @createdAt
  defaultFreeTrialDaysForItem: Int
  globalEndFreeTrialPeriod: DateTime
  globalStartFreeTrialPeriod: DateTime
  millionSearches: Int
  numConnections: Int
  type: String!
  updatedAt: DateTime! @updatedAt
}

type Invoice {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"Invoice_id_seq\\", initialValue: 1, allocationSize: 1)
  amount: Float!
  amountPaid: Float!
  billingAccountId: BillingAccount!
  createdAt: DateTime! @createdAt
  endDate: DateTime!
  invoiceBreakdown: [InvoiceBreakdown]
  paymentMethod: String!
  paymentStatus: String!
  startDate: DateTime!
  updatedAt: DateTime! @updatedAt
}

type InvoiceBreakdown {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"InvoiceBreakdown_id_seq\\", initialValue: 1, allocationSize: 1)
  amount: Float!
  createdAt: DateTime! @createdAt
  endDate: DateTime!
  invoiceId: Invoice!
  itemBillingAccountId: ItemBillingAccount!
  startDate: DateTime!
  updatedAt: DateTime! @updatedAt
}

type Item {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"Item_id_seq\\", initialValue: 1, allocationSize: 1)
  contractId: Contract!
  cost: [Cost]
  createdAt: DateTime! @createdAt
  enabled: Boolean @default(value: false)
  itemBillingAccount: [ItemBillingAccount]
  product: [Product]
  specificFreeTrialDays: Int
  updatedAt: DateTime! @updatedAt
}

type ItemBillingAccount @db(name: \\"Item_BillingAccount\\") {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"Item_BillingAccount_id_seq\\", initialValue: 1, allocationSize: 1)
  billingAccountId: BillingAccount!
  createdAt: DateTime! @createdAt
  invoiceBreakdown: [InvoiceBreakdown]
  itemId: Item!
  updatedAt: DateTime! @updatedAt
}

type Product {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"Product_id_seq\\", initialValue: 1, allocationSize: 1)
  condition: [Condition]
  createdAt: DateTime! @createdAt
  itemId: Item!
  system: String!
  type: String!
  updatedAt: DateTime! @updatedAt
}"
`;
